lib/Dialect/Ibis/Transforms/IbisContainersToHW.cpp

```diff
@@ -0,0 +1,306 @@
+//===- IbisContainersToHW.cpp ---------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/Ibis/IbisDialect.h"
+#include "circt/Dialect/Ibis/IbisOps.h"
+#include "circt/Dialect/Ibis/IbisPasses.h"
+#include "circt/Dialect/Ibis/IbisTypes.h"
+
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace circt;
+using namespace ibis;
+
+static llvm::StringRef getScopeRefModuleName(Type type) {
+  return type.cast<ScopeRefType>().getScopeRef().getAttr().strref();
+}
+
+namespace {
+
+// Analysis result for generating the port interface of a container + a bit of
+// port op caching.
+struct ContainerPortInfo {
+  std::unique_ptr<hw::ModulePortInfo> hwPorts;
+
+  // A mapping between the port name and the port op within the container.
+  llvm::DenseMap<StringAttr, InputPortOp> opInputs;
+
+  // A mapping between the port name and the port op within the container.
+  llvm::DenseMap<StringAttr, OutputPortOp> opOutputs;
+
+  ContainerPortInfo() = default;
+  ContainerPortInfo(ContainerOp container) {
+    SmallVector<hw::PortInfo, 4> inputs, outputs;
+
+    // Gather in and output port ops.
+    for (auto input : container.getBodyBlock()->getOps<InputPortOp>()) {
+      opInputs[input.getSymNameAttr()] = input;
+
+      hw::PortInfo portInfo;
+      portInfo.name = input.getSymNameAttr();
+      portInfo.type = cast<PortOpInterface>(input.getOperation()).getPortType();
+      portInfo.dir = hw::ModulePort::Direction::Input;
+      inputs.push_back(portInfo);
+    }
+
+    for (auto output : container.getBodyBlock()->getOps<OutputPortOp>()) {
+      opOutputs[output.getSymNameAttr()] = output;
+
+      hw::PortInfo portInfo;
+      portInfo.name = output.getSymNameAttr();
+      portInfo.type =
+          cast<PortOpInterface>(output.getOperation()).getPortType();
+      portInfo.dir = hw::ModulePort::Direction::Output;
+      outputs.push_back(portInfo);
+    }
+
+    // Sort the in- and output port lists by name to ensure deterministic
+    // ordering.
+    auto cmp = [](const hw::PortInfo *lhs, const hw::PortInfo *rhs) {
+      return lhs->name.getValue().compare(rhs->name.getValue());
+    };
+    llvm::array_pod_sort(inputs.begin(), inputs.end(), cmp);
```

Comment:
Why use this over just `sort`?

Reply:
Why are you sorting at all? The order could be (will be) meaningful.

Reply:
Fair point, i suspect this code got written with a map in mind, but in reality (as you mention) the order of ops within a block is obviously deterministic, and that determinism in carried through to the `hw::ModulePortInfo`. Will remove.